src/Tools/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/SolutionState.cs

```diff
@@ -0,0 +1,259 @@
+        /// <item><description><c>{|ID1:text|}</c>: indicates that a diagnostic with ID <c>ID1</c> is reported for
+        /// <c>text</c>. The diagnostic descriptor for <c>ID1</c> is located via
+        /// <see cref="AnalyzerTest{TVerifier}.GetDiagnosticAnalyzers"/>. If no matching descriptor is found, the
+        /// diagnostic is assumed to be a compiler-reported diagnostic with the specified ID and severity
+        /// <see cref="DiagnosticSeverity.Error"/>.</description></item>
+        /// </list>
+        /// </remarks>
+        public MarkupMode AllowMarkup { get; set; }
+
+        /// <summary>
+        /// Applies the <see cref="InheritanceMode"/> using a specified base state.
+        /// </summary>
+        /// <remarks>
+        /// <para>This method evaluates <see cref="AdditionalFilesFactories"/>, and places the resulting additional
+        /// files in the <see cref="AdditionalFiles"/> collection of the result before returning.</para>
+        /// </remarks>
+        /// <param name="baseState">The base state to inherit from, or <see langword="null"/> if the current state is
+        /// the root state.</param>
+        /// <param name="fixableDiagnostics">The set of diagnostic IDs to treat as fixable. Fixable diagnostics present
+        /// in the <see cref="ExpectedDiagnostics"/> collection of the base state are never inherited.</param>
+        /// <returns>A new <see cref="SolutionState"/> representing the current state with inherited values applied
+        /// where appropriate. The <see cref="InheritanceMode"/> of the result is
+        /// <see cref="StateInheritanceMode.Explicit"/>.</returns>
+        public SolutionState ApplyInheritedState(SolutionState baseState, ImmutableArray<string> fixableDiagnostics)
+        {
+            Debug.Assert(
+                InheritanceMode == StateInheritanceMode.AutoInherit || InheritanceMode == StateInheritanceMode.Explicit,
+                $"Unexpected inheritance mode: {InheritanceMode}");
+
+            if (baseState?.AdditionalFilesFactories.Count > 0)
+            {
+                throw new InvalidOperationException("The base state should already have its inheritance state evaluated prior to its use as a base state.");
+            }
+
+            var result = new SolutionState(_defaultPrefix, _defaultExtension, AllowMarkup);
+            if (InheritanceMode == StateInheritanceMode.AutoInherit && baseState != null)
+            {
+                if (Sources.Count == 0)
+                {
+                    result.Sources.AddRange(baseState.Sources);
+                }
+
+                if (AdditionalFiles.Count == 0)
+                {
+                    result.AdditionalFiles.AddRange(baseState.AdditionalFiles);
+                }
+
+                if (ExpectedDiagnostics.Count == 0)
+                {
+                    result.ExpectedDiagnostics.AddRange(baseState.ExpectedDiagnostics.Where(diagnostic => !fixableDiagnostics.Contains(diagnostic.Id)));
```

Comment:
📝 Diagnostics inherited *as markup* through `Sources` or `AdditionalFiles` will adhere to the `AllowMarkup` setting of the current instance when `WithProcessedMarkup` is called. However, diagnostics inherited directly from the `ExpectedDiagnostics` collection are always filtered during the inheritance process.